File: KataSix/KataSix.py
class AnagramDictionary(object):
    def __init__(self, filename):
        self.underlying = {}
        self.longest_length = 0
        self.longest_canonical = set()
        self.most_length = 0
        self.most_canonical = set()
        with open(filename) as f:
            for word in f:
                self.insert(word.strip())

    @staticmethod
    def canonical_representation(string):
        return ''.join(sorted(string.lower()))

    def insert(self, string):
        canonical = self.canonical_representation(string)
        anagrams = self.underlying.get(canonical, set())
        anagrams.add(string.lower())
        self.underlying[canonical] = anagrams
        self.update_stats(canonical, anagrams)

    def update_stats(self, canonical, anagrams):
        if len(anagrams) > 1:
            if len(canonical) == self.longest_length:
                self.longest_canonical.add(canonical)
            elif len(canonical) > self.longest_length:
                self.longest_length = len(canonical)
                self.longest_canonical = {canonical}
            if len(anagrams) == self.most_length:
                self.most_canonical.add(canonical)
            elif len(anagrams) > self.most_length:
                self.most_length = len(anagrams)
                self.most_canonical = {canonical}

    def get_anagrams(self, string):
        canonical = self.canonical_representation(string)
        anagrams = self.underlying.get(canonical, set())
        return anagrams - {string.lower()}

    def get_longest_anagrams(self):
        return [self.underlying.get(canonical) for canonical in self.longest_canonical]

    def get_most_anagrams(self):
        return [self.underlying.get(canonical) for canonical in self.most_canonical]

File: ruby/06/KataSix.rb
class AnagramDictionary
  def initialize(filename)
    @underlying = {}
    @longest_length = 0
    @longest_canonical = {}
    @most_length = 0
    @most_canonical = {}
    File.foreach(filename) { |word| insert(word.strip) }
  end

  def self.canonical_representation(string)
    string.downcase.chars.sort.join
  end

  def insert(string)
    canonical = self.class.canonical_representation(string)
    anagrams = @underlying[canonical] ||= {}
    anagrams[string.downcase] = nil
    @underlying[canonical] = anagrams
    update_stats(canonical, anagrams)
  end

  def update_stats(canonical, anagrams)
    if anagrams.size > 1
      if canonical.size == @longest_length
        @longest_canonical[canonical] = nil
      elsif canonical.size > @longest_length
        @longest_length = canonical.size
        @longest_canonical = {canonical => nil}
      end
      if anagrams.size == @most_length
        @most_canonical[canonical] = nil
      elsif anagrams.size > @most_length
        @most_length = anagrams.size
        @most_canonical = {canonical => nil}
      end
    end
  end

  def anagrams(string)
    canonical = self.class.canonical_representation(string)
    anagrams = @underlying[canonical] ||= {}
    anagrams.reject { |k, v| k == string.downcase }.keys
  end

  def longest_anagrams
    @longest_canonical.keys.map { |x| @underlying[x].keys }
  end

  def most_anagrams
    @most_canonical.keys.map { |x| @underlying[x].keys }
  end
end
